[REFACTOR] Cgroup Interface (cgroupv1 and cgroupv2 initialization)

rafaeldtinoco opened the pull request:
## Description (git log)

This is a refactor triggered by comment https://github.com/aquasecurity/tracee/pull/2200#discussion_r990781388. Tracee needs to always initialize cgroupv1 and/or groupv2 when needed. Cgroupv2 is always needed for cgroup ebpf programs.

## Type of change

- [ ] Bug fix (non-breaking change fixing an issue, preferable).
- [ ] Quick fix (minor non-breaking change requiring no issue, use with care)
- [x] Code refactor (code improvement and/or code removal)
- [x] New feature (non-breaking change adding functionality).
- [ ] Breaking change (cause existing functionality not to work as expected).

## How Has This Been Tested?

```
sudo ./dist/tracee-ebpf --install-path /tmp/tracee --cache cache-type=mem --cache mem-cache-size=512 --output format:json --output option:parse-arguments --output option:detect-syscall --containers --trace container --trace comm=bash --trace follow --capabilities bypass=false | jq -c
```

Note: Set bypass to false so capabilities are applied:

```
# cat /proc/$(pidof tracee-ebpf)/status | grep Cap
CapInh: 0000000000000000
CapPrm: 000001ffffffffff
CapEff: 0000000000000000
CapBnd: 0000000000000000
CapAmb: 0000000000000000
```

rafaeldtinoco commented:
@NDStrahilevitz enrichment is still failing (I missed something but likely easy to fix) but I would like your opinion in this change (about the refactoring), considering comments from @yanivagman saying we should mount cgroupv2 (discussion at: https://github.com/aquasecurity/tracee/pull/2200#discussion_r990781388).

NDStrahilevitz commented on pkg/ebpf/tracee.go:
```diff
@@ -328,14 +335,57 @@ func (t *Tracee) Init() error {
+	if err != nil {
+		return err
+	}
+	if !isCgroupV2 {
```
Why should we fail here?
Are we simply dropping support for all non cgroupv1 systems?

NDStrahilevitz commented on pkg/ebpf/tracee.go:
```diff
@@ -328,14 +335,57 @@ func (t *Tracee) Init() error {
+	t.cgroupv2 = cgroup.InitCgroupV2()
+	err = t.cgroupv2.Initialize()
+	if err != nil {
+		return err // cgroupv2 is unavailable (it needs to be for cgroup progs)
```
I don't think we should fail here if we don't use the cgroup progs. It's one thing not to support network events on those systems, another entirely to just drop tracee as a whole.

NDStrahilevitz commented on pkg/ebpf/tracee.go:
```diff
@@ -1251,6 +1322,20 @@ func (t *Tracee) Close() {
+	if t.cgroupv1 != nil {
+		err := t.cgroupv1.Destroy()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "failed to clean cgroupv1: %s", err)
```
There's a logger now, let's use it for new code. If I missed other prints in the code it goes for that as well.

rafaeldtinoco replied:
Yep, I had doubts if I should just start using it for all the fprintfs around when refactoring, I believe we should as well.

rafaeldtinoco replied:
I would like this change to be made for all fprintfs and outside the scope of this PR.

NDStrahilevitz commented on pkg/ebpf/tracee.go:
```diff
@@ -328,14 +335,57 @@ func (t *Tracee) Init() error {
+	if err != nil {
+		return err
+	}
+	t.cgroupv1 = nil
```
No need, this should start as nil by default.

rafaeldtinoco replied:
I wanted to make sure (when testing =D).

NDStrahilevitz commented on pkg/ebpf/tracee.go:
```diff
@@ -328,14 +335,57 @@ func (t *Tracee) Init() error {
+	t.containers, err = containers.New(
+		defaultCgroup, // cgroupv1 or cgroupv2
+		t.config.Sockets,
+		"containers_map",
```
I think there's a defined const for this now once you rebase on main.

rafaeldtinoco replied:
Can't find...

NDStrahilevitz commented on pkg/cgroup/cgroup.go:
```diff
@@ -13,36 +13,167 @@ import (
+		return fmt.Errorf("cgroup: failed to get absolute path: %s", path)
+	}
+	mp, _ := mount.IsMountpointOf(mountPath, CgroupV1FsType)
+	if mp {
```
If we're already mounted I think we can just return successful no?

rafaeldtinoco replied:
It should be mounted only once per temporary tmp directory (with no recurrent callings IMO). Lets see the final version and we can change this if needed.

NDStrahilevitz commented on pkg/cgroup/cgroup.go:
```diff
@@ -13,36 +13,167 @@ import (
+		return fmt.Errorf("cgroup: failed to get absolute path: %s", path)
+	}
+	mp, _ := mount.IsMountpointOf(mountPath, CgroupV2FsType)
+	if mp {
```
ditto

NDStrahilevitz commented on pkg/cgroup/cgroup.go:
```diff
@@ -80,63 +211,81 @@ func GetCgroupPath(rootDir string, cgroupId uint64, subPath string) (string, err
-	}
-	return false, nil
-}
+// // UmountCgroupV1Cpuset umounts cpuset cgroup controller from a temporary place
```
Needs cleanup I assume.

NDStrahilevitz commented on pkg/ebpf/probes/probes.go:
```diff
@@ -2,48 +2,17 @@ package probes
 
```
Love the split, thanks for taking care of that.

NDStrahilevitz reviewed (commented):
Awesome refactor Rafael, I really like the new cgroup package split.
Left some questions and notes, specifically the cgroupv2 requirement concerns me since this complete drop of support may leave some customers in a problem (this is relevant to the networking PR as well since we should fail the probes on unsupported systems).

rafaeldtinoco commented:
> I do think it would be a good idea to just merge this before the networking PR, cherry-pick them on top of main, and then rebase the network PR, because while I tried going through it commit by commit, it would be easier if I had it clean from the netowrk PR diffs.

Yes, I know it got polluted, as the network code is a bit big. Will work in merging this one first (right on it).

rafaeldtinoco commented:
The test is failing in RHEL8 because I believe we currently have a bug at the capabilities logic.

So we are keeping CAP_SYS_ADMIN in NEW kernels (shouldn't it be dropped ?):

![image](https://user-images.githubusercontent.com/7395852/195763263-8981d145-6b8f-4fe1-a428-d8faf2eab691.png)

AND for OLD kernels (that seems correct):

![image](https://user-images.githubusercontent.com/7395852/195763282-a98e809c-53b3-4d69-b06b-956c71bac5f0.png)

and dropping it for RHEL8 only (this is correct, but by accident):

![image](https://user-images.githubusercontent.com/7395852/195763313-7623bc3c-59bb-4eba-ac1e-f442cc593381.png)

I believe the capabilities feature is currently broken.

rafaeldtinoco commented:
This PR is held until https://github.com/aquasecurity/tracee/pull/2243 is merged.

rafaeldtinoco commented:
This PR is built on top of https://github.com/aquasecurity/tracee/pull/2243 (which needs to be merged first). That PR takes care of the "new capabilities" feature (privilege rings) and this one uses it for mounting and unmounting cgroupv1 and cgroupv2 during tracee startup/shutdown time.

rafaeldtinoco commented:
This should be reviewed before https://github.com/aquasecurity/tracee/pull/2200 because that PR needs this one first (and contains the same commits for the time being).

NDStrahilevitz commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,292 @@
+)
+
+type Cgroup interface {
+	initialize() error
```
Since interfaces deal with what is publicly available, it's a bit unnecessary to have `initalize` here, and i'll expand further about something that may be simpler to do.

NDStrahilevitz commented on pkg/containers/containers.go:
```diff
@@ -15,19 +15,15 @@ import (
 type Containers struct {
-	cgroupV1   bool
-	cgroupMP   string
+	defCgroup  cgroup.Cgroup
```
nit: think just `cgroup` is enough

NDStrahilevitz commented on pkg/containers/containers.go:
```diff
@@ -46,20 +42,23 @@ type CgroupInfo struct {
 // User should further call "Populate" and iterate with Containers data.
-func New(sockets cruntime.Sockets, mapName string, debug bool) (*Containers, error) {
+func New(
+	defCgroup cgroup.Cgroup,
```
ditto, the "default" concept should be irrelevant for the reading here (and we can make the "default" logic a bit simpler IMO i'll expand upon it further in the review).

NDStrahilevitz commented on pkg/ebpf/tracee.go:
```diff
@@ -408,17 +413,62 @@ func (t *Tracee) Init() error {
+	var defaultCgroup cgroup.Cgroup
 
-		t.containers, err = containers.New(t.config.Sockets, "containers_map", t.config.Debug)
+	err = caps.Requested(func() error { // ring2
```
So I think all the nested logic here can go into a function in the new `cgroup` package of `cgroup.New()` that will return an instance of the `Cgroup` interface, that way we may be able to not have two instances of cgroup at the same time on tracee's struct level.
It might be necessary to add a "combined" v1+v2 cgroup struct impl for this, in which case it may be better to just leave it be.

NDStrahilevitz commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,292 @@
+
+// cgroupv1
+
+func InitCgroupV1() (Cgroup, error) {
```
So I think it would be best if we had here some `func New() (Cgroup, error)` and it would be responsible for decision making of which cgroup implementation to return, and that would be the preferred way to construct a cgroup object.

NDStrahilevitz replied:
That way in tracee.go itself there will be minimal exposure of internal decision making of which cgroups we need, just `t.cgroup, err := cgroup.New()` (obviously inside a `capabilities.Required` callback).

NDStrahilevitz commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,292 @@
+	Destroy() error
+	GetMountPoint() string
+	GetDefaultHierarchy() int
+	IsV1() bool
```
If we go for the approach I suggested, we may need to also add `IsV2() bool` for the combined implementation (if it's a necessary one).

NDStrahilevitz reviewed (approved):
I don't see any obvious issues here, only substantial comments I have are about some naming and perhaps a different "construction" structure for the cgroup module, up to you if you find it better, but LGTM for merging, not going to block on style.

yanivagman commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,292 @@
+
+	c.mountPath = mountPath
+
+	c.hierarchy, err = GetCgroupControllerHierarchy("cpuset")
```
Let's not make the hierarchy "cpuset" hardcoded

yanivagman commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,292 @@
+	hierarchy int
+}
+
+func (c *cgroupV2) initialize() error {
```
This function looks the same as the cgroup v1 initialize function. I think it can be generalized to take relevant parameters and handle both to avoid code duplication

yanivagman commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,292 @@
+
+	c.mountPath = mountPath
+
+	c.hierarchy, err = GetCgroupControllerHierarchy("cpuset")
```
Let's avoid hardcoding "cpuset"

yanivagman commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,292 @@
+	return nil
+}
+
+func (c *cgroupV2) Destroy() error {
```
Again, this function is unnecessarily duplicated of cgroup v1

yanivagman commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,292 @@
+	cpuset, err[1] = GetCgroupControllerHierarchy("cpuset")
+	memory, err[2] = GetCgroupControllerHierarchy("memory")
+
+	for _, e := range err {
```
It can be (although less common) that not all cgroup hierarchies are compiled into the kernel, but here you assume they always are.

rafaeldtinoco replied:
Yes, but I'm assuming the 3 most important ones are always available. I believe they are in all environments we currently test tracee,

rafaeldtinoco replied:
@yanivagman, I'm reworking this PR entirely, after refactoring the code in other already merged PRs, but, about:


> I think we should opt to use the previous method for deciding about group v1/v2 since it is the standard way to do this check and also runc uses it (and we need to be consistent with runc): https://github.com/opencontainers/runc/blob/main/libcontainer/cgroups/fs2/create.go#L14

> This is important for correctly resolving the container ID from the cgroup path. Also, it will allow not to return an error from this function, and if the file doesn't exist we can simply assume cgroup v1 (as we did with the current implementation)

About hierarchyID, the cgroups man page states:

```
    2. The unique ID of the cgroup hierarchy on which this
       controller is mounted.  If multiple cgroups v1
       controllers are bound to the same hierarchy, then each
       will show the same hierarchy ID in this field.  The
       value in this field will be 0 if:

       a) the controller is not mounted on a cgroups v1
          hierarchy;

       b) the controller is bound to the cgroups v2 single
          unified hierarchy; or

       c) the controller is disabled (see below).
```

So, the previous method (currently in use) always assumes that cgroupv1 is mounted in /sys/fs/cgroup tmpfs and that might not always be the case. We could have unmounted cgroupv1 but, still, being the default, for example.

Kubernetes documentation (https://kubernetes.io/docs/concepts/architecture/cgroups/) is also misleading as they tell user to identify if "/sys/fs/cgroup" is a tmpfs or a cgroup2 filesystem (but in our case the filesystems might not even be mounted).

With that said, I believe the correct way of detecting the default cgroup version being used is through hiearchyID of cpuset from /proc/cgroups.

rafaeldtinoco replied:
About last comment, reading: https://github.com/systemd/systemd/blob/69c0807432fa4fbfbf507a53872664cd26715559/src/basic/cgroup-util.c#L1986 we see that systemd relies in executing stat on "/sys/fs/cgroup" to check if it is a cgroup2 filesystem (same approach documented by kubernetes). The fact is that we might run tracee inside a container with no group mounted (let's say). Then, the /proc/cgroups is the best alternative to understand which group is the default and to mount one (or both) cgroup versions.

yanivagman reviewed (changes requested):
I will finish reviewing this one later, but wanted to post some comments I already have

rafaeldtinoco commented:
@NDStrahilevitz could you check the changes I made ? I believe I have addressed what you suggested. I preferred to use a "Cgroups" object to contain both "Cgroup" objects (cgroupv1 and cgroupv2). This way the initialization from the caller perspective is easier (and the cleanup as well).

Let me know if you have any other comments pls.

Note: I believe I address all Yaniv's comments in this version as well (mostly were about code duplication, which was removed in this new commit).

josedonizetti commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,312 @@
+		}
+	}
+
+	if cpu == cpuset && cpu == memory {
```
@rafaeldtinoco how can `cpu` be `cpuset` and `memory` at the same time? 
```suggestion
	if cpu == cpuset && cpu == memory {
```

rafaeldtinoco replied:
Because in cgroupv2 they are all the same Id

josedonizetti replied:
Ahhh! Maybe add a comment about it?

josedonizetti commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,312 @@
+// Constants
+//
+
+const procCgroups = "/proc/cgroups"
+const procFilesystems = "/proc/filesystems"
+
+const CgroupV1FsType = "cgroup"
+const CgroupV2FsType = "cgroup2"
+
+const DefaultCgroupController = "cpuset"
+
+const TmpCgroup = "/tmp"
+const TmpCgroupV1Path = "tracee-cgroupv1"
+const TmpCgroupV2Path = "tracee-cgroupv2"
```
Nit

```suggestion
const (
  procCgroups = "/proc/cgroups"
  procFilesystems = "/proc/filesystems"
  
  CgroupV1FsType = "cgroup"
  CgroupV2FsType = "cgroup2"
  
  DefaultCgroupController = "cpuset"
  
  TmpCgroup = "/tmp"
  TmpCgroupV1Path = "tracee-cgroupv1"
  TmpCgroupV2Path = "tracee-cgroupv2"
)
```

josedonizetti commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,312 @@
+
+	isCgroupV1, err := IsCgroupV1Default() // initialize cgroupv1
+	if err != nil {
+		return cs, err
```
considering we are returning a pointer, best to return nil if there is an error, no?
if you agree, it would make sense to change all returns of errors on this method

```suggestion
		return nil, err
```

rafaeldtinoco replied:
We tend to return an empty object in some other places... and I had the same doubt when I did it. Thinking a bit more, after your comment, since I'm setting t.cgroups directly, maybe it is better to have it as nil if not initialized correctly.

josedonizetti commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,312 @@
+// Destroy unmounts cgroupfs that were temporarily mounted
+func (cs *Cgroups) Destroy() error {
+	if cs.cgroupv1 != nil {
+		err := cs.cgroupv1.Destroy()
```
nit
```suggestion
		return cs.cgroupv1.Destroy()
```

rafaeldtinoco replied:
Then I would not destroy cgroupv2 if cgroupv1 does not exist (I might have cgroupv1 + cgroupv2 or cgroupv2 only).

josedonizetti commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,312 @@
+		}
+	}
+	if cs.cgroupv2 != nil {
+		err := cs.cgroupv2.Destroy()
```
nit

```suggestion
		return cs.cgroupv2.Destroy()
```

rafaeldtinoco replied:
I could indeed return here directly, preferred not to just "because" (so both calls were the same =D).

josedonizetti commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,312 @@
+		fsType = CgroupV2FsType
+		controller = "" // no default controller for cgroupv2
+	default:
+		return c, failWrongCgroupVersion()
```
considering we are returning a pointer, best to return nil if there is an error, no?
if you agree, it would make sense to change all returns of errors on this method

```suggestion
		return nil, failWrongCgroupVersion()
```

josedonizetti commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,312 @@
+	if err != nil {
+		return err
+	}
+	err = os.RemoveAll(c.mountPath)
```
nit
```suggestion
	return os.RemoveAll(c.mountPath)
```

josedonizetti commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,312 @@
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for i := 1; scanner.Scan(); i++ {
```
Nit
```suggestion
	for scanner.Scan() {
```

josedonizetti commented on pkg/mount/mount.go:
```diff
@@ -0,0 +1,55 @@
+
+	mp := ""
+	scanner := bufio.NewScanner(file)
+	for i := 1; scanner.Scan(); i++ {
```
nit
```suggestion
	for scanner.Scan() {
```

josedonizetti commented on pkg/mount/mount.go:
```diff
@@ -0,0 +1,55 @@
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for i := 1; scanner.Scan(); i++ {
```
nit
```suggestion
	for scanner.Scan() {
```

josedonizetti commented on pkg/ebpf/probes/common.go:
```diff
@@ -0,0 +1,23 @@
+
+// enableDisableAutoload enables or disables an eBPF program autoload setting
+func enableDisableAutoload(module *bpf.Module, programName string, autoload bool) error {
+	var err error
```
nit, we can remove this line

josedonizetti reviewed (commented):
@rafaeldtinoco Added a question about an if that seems would never work, and a couple of nits, feel free to ignore them if you don't agree.

rafaeldtinoco commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,312 @@
+		fsType = CgroupV2FsType
+		controller = "" // no default controller for cgroupv2
+	default:
+		return c, failWrongCgroupVersion()
```
Yes, same case as the other one. I agree.

rafaeldtinoco commented:
@josedonizetti thanks for the review Jose, I have addressed most of your nits and suggestions and pushed a new commit. I'm using new commits because I'm rebasing the new-network PR on top of this one.

rafaeldtinoco commented:
Okay @yanivagman, I guess it is on you now. Let me know if you have any suggestions/concerns, please. Thanks.

yanivagman commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,311 @@
+//
+
+// Cgroups handles cgroups initialization (v1+v2 or v2 only)
+type Cgroups struct {
+	cgroupv1 *Cgroup // used for cgroup filesystem readings
+	cgroupv2 *Cgroup // used for cgroup prog attachment (and readings, if v2 is default)
+	cgroup   *Cgroup // default cgroup (might point to one or another)
+}
```
Why do we need to keep both cgroupv1 and cgroupv2? Won't having the deafult cgroup be enough?

NDStrahilevitz replied:
I think this one takes care of the case where you have a hybrid system with both.

rafaeldtinoco replied:
I believe this is solved (Nadav explained the reason).

yanivagman commented on pkg/ebpf/tracee.go:
```diff
@@ -408,17 +412,31 @@ func (t *Tracee) Init() error {
+		return err
+	}, cap.SYS_ADMIN)
+	if err != nil {
+		return err
```
This will fail tracee from running. But is that what we really want? What if the user don't need cgroup programs?

NDStrahilevitz replied:
We need the mounting logic anyway when we run tracee inside a container (see from a while back #1338 #1900)

yanivagman replied:
But now it is always required, which makes the requirement for CAP_BPF and CAP_PERFMON on newer kernels irrelevant.
CAP_SYS_ADMIN is something we should avoid. We need to have the ability to let the user run tracee without requiring this capability - either by running the binary directly or by mounting the cgroup themselves.

NDStrahilevitz replied:
We only need to run this in a capabilities ring once during initialization and that's it, other than that we can keep running only with CAP_BPF and CAP_PERFMON.
I do think we should move this caps call to the cgroups method though, that way we can scope it only for the mount call itself.

yanivagman replied:
The thing is that there are environments where CAP_SYS_ADMIN won't be available for us, and we need to try to run tracee even without it (of course, if we can't continue - we can return an error)

rafaeldtinoco replied:
I'll do like what we've discussed and mount only when needed. For mounting, I'll raise to ring2 with CAP_SYSADMIN only (and will have to unmounted at the cleanup).

rafaeldtinoco replied:
This is being solved by new code.

yanivagman commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,311 @@
+	case CgroupV2:
+		tmpPath = TmpCgroupV2Path
+		fsType = CgroupV2FsType
+		controller = "" // no default controller for cgroupv2
```
nit: No need to initialize the variable to an empty string. This is the default value of a string variable.

yanivagman commented on pkg/containers/containers.go:
```diff
@@ -229,12 +138,12 @@ func (c *Containers) populate() error {
-// saving container information in Containers CgroupInfo map.
-// NOTE: ALL given cgroup dir paths are stored in CgroupInfo map.
+// saving container information in Containers CgroupInfo map.  NOTE: ALL given
+// cgroup dir paths are stored in CgroupInfo map.
```
This doesn't make the comment more readable

yanivagman reviewed (changes requested):
Ok, I finished reviewing this one.
I like the refactor you've done, but there are a few points that might cause tracee to work on less environments. Please see my comments below.

Also, I think it would have been easier if you would have made one refactor PR (with no logic changes at all), and another one with your new suggested changes. We can merge the refactor quickly, and then discuss the required changes for the cgroup stuff. Your call if you want to do this split or not.

rafaeldtinoco commented:
@yanivagman, I'm reworking this PR entirely, after refactoring the code in other already merged PRs, but, about:


> I think we should opt to use the previous method for deciding about group v1/v2 since it is the standard way to do this check and also runc uses it (and we need to be consistent with runc): https://github.com/opencontainers/runc/blob/main/libcontainer/cgroups/fs2/create.go#L14

> This is important for correctly resolving the container ID from the cgroup path. Also, it will allow not to return an error from this function, and if the file doesn't exist we can simply assume cgroup v1 (as we did with the current implementation)

About hierarchyID, the cgroups man page states:

```
    2. The unique ID of the cgroup hierarchy on which this
       controller is mounted.  If multiple cgroups v1
       controllers are bound to the same hierarchy, then each
       will show the same hierarchy ID in this field.  The
       value in this field will be 0 if:

       a) the controller is not mounted on a cgroups v1
          hierarchy;

       b) the controller is bound to the cgroups v2 single
          unified hierarchy; or

       c) the controller is disabled (see below).
```

So, the previous method (currently in use) always assumes that cgroupv1 is mounted in /sys/fs/cgroup tmpfs and that might not always be the case. We could have unmounted cgroupv1 but, still, being the default, for example.

Kubernetes documentation (https://kubernetes.io/docs/concepts/architecture/cgroups/) is also misleading as they tell user to identify if "/sys/fs/cgroup" is a tmpfs or a cgroup2 filesystem (but in our case the filesystems might not even be mounted).

With that said, I believe the correct way of detecting the default cgroup version being used is through hiearchyID of cpuset from /proc/cgroups.

rafaeldtinoco commented:
About last comment, reading: https://github.com/systemd/systemd/blob/69c0807432fa4fbfbf507a53872664cd26715559/src/basic/cgroup-util.c#L1986 we see that systemd relies in executing stat on "/sys/fs/cgroup" to check if it is a cgroup2 filesystem (same approach documented by kubernetes). The fact is that we might run tracee inside a container with no group mounted (let's say). Then, the /proc/cgroups is the best alternative to understand which group is the default and to mount one (or both) cgroup versions.

rafaeldtinoco commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,311 @@
+
+	DefaultCgroupController = "cpuset"
+
+	TmpCgroup       = "/tmp"
```
Actually this brings to a good topic... we should use default OS temp directory for temporary files (like these and PID files) AND, perhaps, rename "tmpdir" to something else (it is not a temporary directory, it is an artifacts output directory). No ?

rafaeldtinoco replied:
Also, if we want to use a "global tmpdir", then we should have global variable to it (instead of having it as part of Tracee object). Or else we will have to pass it as an argument for all packages using it.

rafaeldtinoco commented on pkg/cgroup/cgroup.go:
```diff
@@ -0,0 +1,311 @@
+	// cgroupv1: mount -t cgroup -o cpuset cgroup mountPath
+	// cgroupv2: mount -t cgroup2 cgroup2 mountPath
+
+	err = syscall.Mount(fsType, mountPath, fsType, 0, controller) // mount
```
Being solved by new code.

NDStrahilevitz commented on pkg/cgroup/cgroup.go:
```diff
@@ -14,73 +12,332 @@ import (
+
+func NewCgroup(ver CgroupVersion) (Cgroup, error) {
+	switch ver {
+	case CgroupVersion1:
```
this can be shorter:
```
var c Cgroup
switch ver {
case CgroupVersion1:
     c = &Cgroup
...
}
err := c.init()
return c, err
```

NDStrahilevitz commented on pkg/mount/mount.go:
```diff
@@ -2,54 +2,214 @@ package mount
+// General
+//
+
+func IsDirEmpty(pathname string) (bool, error) {
```
I think this can move to `pkg/utils/files.go`

NDStrahilevitz commented on pkg/containers/containers.go:
```diff
@@ -17,16 +17,12 @@ import (
-	cgroupMP   string
-	cgroups    map[uint32]CgroupInfo
+	cgroups    *cgroup.Cgroups
+	cgroupsMap map[uint32]CgroupInfo
```
suggestion: maybe `cgroupsInfo` or `cgroupsCache`? WDYT?

rafaeldtinoco replied:
Nope, Info is already used and it is not a cache =).

yanivagman commented on pkg/mount/mount.go:
```diff
@@ -2,54 +2,214 @@ package mount
+// private
+
+func (m *MountOnce) isMountedByOS() (bool, error) {
+	mp, err := SearchMountpoint(m.fsType, m.data)
```
I would only look at standard mountpoints (e.g. /sys/fs/cgroup or /tmp/tracee/*) and not on any mountpoint.
It can be that some other userspace program has mounted cgroup fs (e.g. Cilium), and we will find it first. But then, if this other program stops running, it might also unmount the fs.

rafaeldtinoco replied:
Fair enough.

rafaeldtinoco commented on pkg/cgroup/cgroup.go:
```diff
@@ -14,73 +12,332 @@ import (
+	CgroupVersionNone CgroupVersion = 1 << iota
+	CgroupVersion1
+	CgroupVersion2
+	CgroupVersionAll CgroupVersion = CgroupVersion1 & CgroupVersion2
```
Its a leftover from a previous idea, making it iota.

rafaeldtinoco commented on pkg/ebpf/tracee.go:
```diff
@@ -408,17 +410,30 @@ func (t *Tracee) Init() error {
+	// Initialize containers enrichment logic
 
-	}, cap.SYS_ADMIN)
+	capabilities.GetInstance().Requested(
```
No, actually you're right. Nice catch! Thanks

rafaeldtinoco commented on pkg/cgroup/errors.go:
```diff
@@ -0,0 +1,21 @@
+
+import "fmt"
+
+type VersionNotSupported struct{}
```
Because of:

![image](https://user-images.githubusercontent.com/7395852/203098019-cca34902-3f30-4680-8ae2-af9fb4b36e52.png)

I'm actually checking the type of the error to take decisions (unsupported version is an accepted error).

rafaeldtinoco commented on pkg/cgroup/cgroup.go:
```diff
@@ -14,73 +12,332 @@ import (
+
+func NewCgroups() (*Cgroups, error) {
+	var err error
+	var cgrp *Cgroup
```
This is a pointer of a pointer (its the default cgroup, pointing to either cgroupv1 or cgroupv2).

rafaeldtinoco commented on pkg/cgroup/cgroup.go:
```diff
@@ -14,73 +12,332 @@ import (
 		}
-		intID, err := strconv.Atoi(sline[1])
-		if err != nil || intID < 0 {
-			return -1, fmt.Errorf("error parsing %s: %w", cgroupsFile, err)
```
Nothing, returning it.

rafaeldtinoco commented on pkg/cgroup/cgroup.go:
```diff
@@ -14,73 +12,332 @@ import (
+
+	file, err := os.Open(procCgroups)
 	if err != nil {
-		return -1, fmt.Errorf("could not open cgroups file %s: %w", cgroupsFile, err)
```
nothing, returning it.

rafaeldtinoco commented on pkg/cgroup/cgroup.go:
```diff
@@ -14,73 +12,332 @@ import (
+	// discover the default cgroup being used
+	defaultVersion, err := GetCgroupDefaultVersion()
+	if err != nil {
+		return nil, err
```
I believe for not to fail in the inexistence of a cgroupfs we should review containers.go as it might depend on its existence (the change would have to be coordinated among the two). I have opened https://github.com/aquasecurity/tracee/issues/2371 to be checked after this PR is merged (it is blocking the new network code for sometime now).

yanivagman commented on pkg/cgroup/cgroup.go:
```diff
@@ -14,73 +12,352 @@ import (
+
+	ok, err := IsCgroupV1MountedAndDefault()
+	if err != nil {
+		return -1, err
```
maybe we should continue with the second method instead of failing here?

yanivagman commented on pkg/cgroup/cgroup.go:
```diff
@@ -14,73 +12,352 @@ import (
+//
+// https://www.kernel.org/doc/Documentation/cgroup-v1/cgroups.txt
+func IsCgroupV1MountedAndDefault() (bool, error) {
+	_, err := os.Stat(CgroupReleaseAgent)
```
the method we used was to check for the existance of cgroup.controller (this is also used by runc) - shouldn't we continue with that?

rafaeldtinoco replied:
Actually I almost copied the function we had and it was:
```diff
-func RunsOnContainerV1() (bool, error) {
-       _, err := os.Stat(CgroupReleaseAgent)
-       if os.IsNotExist(err) {
-               return true, nil
-       }
-       if err != nil {
-               return false, err
-       }
-       return false, nil
-}
```

rafaeldtinoco replied:
![image](https://user-images.githubusercontent.com/7395852/203159283-4ed62c41-3efb-4723-8cfd-d4a7ead807d8.png)

yanivagman replied:
Like discussed offline, this was used to check if we run in a container or not once we already know that we are on a cgroup v1 environment (by checking for cgroup.controller existance)

yanivagman commented on pkg/mount/mount.go:
```diff
@@ -3,53 +3,202 @@ package mount
+	// Check for mounted cgroupfs in sysfs only (other mountpoints might be
+	// being managed by other softwares, and they could unmount the cgroupfs
+	// unexpectedly)
+	if mp == "" || !strings.Contains(mp, sysFsCgroupPrefix) {
```
This package should be a generic mount package, no?
If so, we should get this as a parameter and not hardcode the cgroup prefix

yanivagman reviewed (approved):
Thanks for addressing all the comments @rafaeldtinoco 
LGTM now
